pkg/apis/kubeone/config/config.go

```diff
@@ -222,15 +224,61 @@ func SetKubeOneClusterDynamicDefaults(cfg *kubeoneapi.KubeOneCluster, credential
 
 	// Source cloud-config from the credentials file if it's present
 	if cc, ok := credentials["cloudConfig"]; ok {
-		cfg.CloudProvider.CloudConfig = cc
+		cluster.CloudProvider.CloudConfig = cc
 	}
 	// Source csi-config from the credentials file if it's present
 	if cc, ok := credentials["csiConfig"]; ok {
-		cfg.CloudProvider.CSIConfig = cc
+		cluster.CloudProvider.CSIConfig = cc
+	}
+
+	if ra, ok := credentials["registriesAuth"]; ok {
+		if err := setRegistriesAuth(cluster, ra); err != nil {
+			return fmt.Errorf("failed to parse registriesAuth from credentials file: %w", err)
+		}
 	}
 
 	// Default the AssetsConfiguration internal API
-	cfg.DefaultAssetConfiguration()
+	cluster.DefaultAssetConfiguration()
+
+	return nil
+}
+
+func setRegistriesAuth(cluster *kubeoneapi.KubeOneCluster, buf string) error {
+	var (
+		containerdConfig kubeonev1beta2.ContainerRuntimeContainerd
+		typeMeta         runtime.TypeMeta
+	)
+
+	if err := yaml.Unmarshal([]byte(buf), &typeMeta); err != nil {
+		return err
+	}
+
+	if typeMeta.APIVersion != kubeonev1beta2.SchemeGroupVersion.String() {
+		return fmt.Errorf("only %q apiVersion is supported in registriesAuth", kubeonev1beta2.SchemeGroupVersion.String())
+	}
+
+	containerdConfigKind := reflect.TypeOf(containerdConfig).Name()
+	if typeMeta.Kind != containerdConfigKind {
+		return fmt.Errorf("only %q kind is supported in registriesAuth", containerdConfigKind)
+	}
+
+	if err := yaml.Unmarshal([]byte(buf), &containerdConfig); err != nil {
+		return err
+	}
+
+	if cluster.ContainerRuntime.Containerd == nil {
+		return fmt.Errorf(".ContainerRuntime.Containerd should be set")
+	}
+
+	if cluster.ContainerRuntime.Containerd.Registries == nil {
+		cluster.ContainerRuntime.Containerd.Registries = map[string]kubeoneapi.ContainerdRegistry{}
+	}
+
+	for registryName, registryInfo := range containerdConfig.Registries {
+		internalRegistry := cluster.ContainerRuntime.Containerd.Registries[registryName]
+		internalRegistry.Auth = (*kubeoneapi.ContainerdRegistryAuthConfig)(registryInfo.Auth)
```

Comment:
If I understand this correctly, this is going to map the v1beta2 representation to the internal representation. While this works fine now, is this going to work if we change the internal API for any reason? Would it be better to have some conversion function that's going to map each field from the versioned to the internal API?

Reply:
It will blow up in compile time if something will be changed in internal cluster API, and that's deliberately! This compile will let us know for sure that something has been changed and we need to take a look and be aware.